### PR TITLE
feat: add progress request batching

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Adds a hook that batches known progress queries into a single request. Uses this hook at the `tutorial-view` level.

## 🤷 Why

To avoid sending many simultaneous requests to the Learn API as we add more `progress` display components.

## 🛠️ How

- Adds `use-has-waited-for-query`, which allows individual progress queries to wait on a single batch progress query
- Uses `use-has-waited-for-query` in `use-tutorial-progress` and `use-collection-progress`
- Adds `use-progress-batch-query` and uses it on `tutorial-view`

## 🧪 Testing

- [ ] Pull this branch down locally
- [ ] [EDIT] Ensure dependencies are up-to-date with `npm i`
    - The work in this PR relies on a specific version of `query-string`
- [ ] Sign in and visit any tutorial view
- [ ] With the `useProgressBatchQuery` in `views/tutorial-view/index.tsx` commented out, you should see a bunch of individual requests to the Learn API in your browser's network tab
- [ ] With the `useProgressBatchQuery` in `views/tutorial-view/index.tsx` active, you should see a single request to the Learn API in your browser's network tab

## 💭 Anything else?

- ✅ EDIT: fixed in ambient zoom! Thanks y'all. ~For some reason I'm seeing `400` requests for `collection-progress` after tutorial progress mutations happen.~
    - I'm not sure if this was happening prior to this PR, I need to double-check.
    - Seems related to `invalidate`? Eventually we make a request for `collection-progress` which does have the token as expected.
    - Ignoring this for now to focus on the core purpose of this PR (batching progress queries!)

[preview]: https://dev-portal-git-zsbatch-fetch-tutorial-progress-hashicorp.vercel.app/
[task]: https://app.asana.com/0/1202097197789424/1202917664989265/f

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -168,6 +168,7 @@
 		"nuka-carousel": "^4.8.4",
 		"p-map": "^4.0.0",
 		"promise-polyfill": "^8.2.1",
+		"query-string": "^7.1.1",
```
Comment (zchsh) [inline package.json:171]: Needed this as an explicit dep to fix an issue in `getAllProgress`

File: src/views/tutorial-view/index.tsx [hunk 2]
```diff
@@ -191,26 +192,50 @@ function TutorialView({
...
+		(t: TutorialLite) => t.id
+	)
+
+	/**
+	 * Prime `tutorial` and `collection` progress queries with a batch query.
+	 *
+	 * This should ideally include all `tutorial` and `collection` entries
+	 * we expect to render on the page, so that we only make one progress request.
+	 */
+	useProgressBatchQuery({
```
Comment (zchsh) [inline src/views/tutorial-view/index.tsx:210]: This view probably merits some cleanup as it's getting chonky....  maybe even some grouping of all `progress` stuff into one hook (`useTutorialViewProgress`)? For now, focus was on "make it work". Very open to ideas here 👍

File: src/hooks/progress/index.ts [hunk 3]
```diff
@@ -1,5 +1,15 @@
-export const TUTORIAL_PROGRESS_QUERY_ID = 'tutorialProgress'
+/**
+ * Constants
```
Comment (zchsh) [inline src/hooks/progress/index.ts:2]: Have added `const`s here so that queries can reliably be aware of each other, which feels important for batching, and for query invalidation.

File: src/hooks/progress/use-collection-progress.ts [hunk 4]
```diff
@@ -23,15 +28,22 @@ interface UseCollectionProgressResult
...
-	// Get the current user's access token
+	/**
+	 * Get the current user's access token, and the batch query status.
+	 * We enable the query only if:
+	 * - an accessToken is present
+	 * - a batch query, if one exists, has been attempted
+	 */
 	const { session } = useAuthentication()
 	const accessToken = session?.accessToken
+	const hasWaitedForBatchQuery = useHasWaitedForQuery([PROGRESS_BATCH_QUERY_ID])
```
Comment (zchsh) [inline src/hooks/progress/use-collection-progress.ts:39]: Have updated `use-collection-progress` to wait on any progress batch query before being enabled (if a batch query exists - if not, we don't wait).

File: src/hooks/progress/use-progress-batch-query/index.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+	// Get the current user's access token
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+
+	/**
+	 * After a successful query, prime the requested
+	 * tutorial & collection progress queries.
+	 */
+	const onSuccess = (data: QueryDataType) => {
```
Comment (zchsh) [inline src/hooks/progress/use-progress-batch-query/index.ts:72]: This batch query is relatively straightforward. There's a bit of nuance around how we process the resulting data, in that we use the original arguments rather than the `data` itself to determine which parts of the cache to update. Have tried to make the reasoning behind this clear in comments.

File: src/hooks/progress/use-tutorial-progress-mutations/index.ts [hunk 6]
```diff
@@ -5,14 +5,16 @@ import {
...
 	TutorialProgressMutationVariables,
 	TutorialProgressMutationArgs,
 	UseTutorialProgressMutationsResult,
 } from './types'
 import { ApiCollectionTutorialProgress } from 'lib/learn-client/api/api-types'
-import { TUTORIAL_PROGRESS_QUERY_ID } from '..'
+import {
+	TUTORIAL_PROGRESS_SINGLE_QUERY_ID,
+	COLLECTION_PROGRESS_SINGLE_QUERY_ID,
+} from '..'
```
Comment (zchsh) [inline src/hooks/progress/use-tutorial-progress-mutations/index.ts:17]: Updated with shared `const` values here.

File: src/hooks/progress/use-tutorial-progress-mutations/index.ts [hunk 7]
```diff
@@ -31,22 +33,25 @@ const useTutorialProgressMutations = (): UseTutorialProgressMutationsResult => {
...
 			queryClient.setQueryData(
-				[TUTORIAL_PROGRESS_QUERY_ID, tutorialId, collectionId],
+				[TUTORIAL_PROGRESS_SINGLE_QUERY_ID, tutorialId, collectionId],
 				data
 			)
 
 			// Invalidate related collection progress
-			queryClient.invalidateQueries(['collectionProgress', collectionId])
+			queryClient.invalidateQueries([
+				COLLECTION_PROGRESS_SINGLE_QUERY_ID,
```
Comment (zchsh) [inline src/hooks/progress/use-tutorial-progress-mutations/index.ts:52]: 🚧 I need to figure out the requests-without-token `400` issue, which I think is related to `invalidateQueries` here... but _might_ try to leave that for a separate PR (I'm not sure how spiky of an issue it is yet).

File: src/hooks/progress/use-tutorial-progress.ts [hunk 8]
```diff
@@ -25,20 +26,30 @@ const useTutorialProgress = ({
...
-	// Get the current user's access token
+	/**
+	 * Get the current user's access token, and the batch query status.
+	 * We enable the query only if:
+	 * - an accessToken is present
+	 * - a batch query, if one exists, has been attempted
+	 */
 	const { session } = useAuthentication()
 	const accessToken = session?.accessToken
+	const hasWaitedForBatchQuery = useHasWaitedForQuery([PROGRESS_BATCH_QUERY_ID])
```
Comment (zchsh) [inline src/hooks/progress/use-tutorial-progress.ts:37]: As with `use-collection-progress`, we wait on any progress batch query (if there is one) before enabling individual tutorial progress queries.

File: src/hooks/use-has-waited-for-query.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * Hook to return whether we've attempted a particular query yet.
+ *
+ * This is intended for use in individual progress queries,
+ * in order to check whether a relevant batch query has been attempted.
+ *
+ * Why? We want to give the batch query a chance to prime the cache
+ * before firing off individual queries. So, we likely want to disable
+ * individual queries until a batch query has been attempted at least once.
+ */
+export function useHasWaitedForQuery(queryKey: QueryKey): boolean {
```
Comment (zchsh) [inline src/hooks/use-has-waited-for-query.ts:13]: This allows to instruct individual tutorial and collection progress queries to wait on a batch progress query (if one has been initialized).

- If there is not a matching batch progress query according to `queryClient.getQueryState`, individual queries will not wait (and we'll see many individual progress requests to the Learn API).
- If there is a matching batch progress query, individual queries will wait till it settles in some way (and we should see only a single batch progress request to the Learn API for progress).

File: src/lib/learn-client/api/progress/get-all-progress.ts [hunk 10]
```diff
@@ -31,9 +31,23 @@ const getAllProgress = async ({
 	tutorialIds,
 	collectionIds,
 }: GetAllProgressOptions): Promise<GetAllProgressResult> => {
+	/**
+	 * TODO: figure out why this happens sometimes!
+	 * Maybe related to "queryClient.invalidate"? That's my best guess so far.
+	 */
+	const doesNotHaveAccessToken = typeof accessToken == 'undefined'
```
Comment (zchsh) [inline src/lib/learn-client/api/progress/get-all-progress.ts:38]: 🚧 Need to figure this out, very open to ideas if anyone has any. Mostly focused on getting the core "batching" part of things sorted first though.

File: src/lib/learn-client/api/progress/get-all-progress.ts [hunk 11]
```diff
@@ -31,9 +31,23 @@ const getAllProgress = async ({
...
+			"getAllProgress: for some reason we're making a request even though we don't have an accessToken."
+		)
+	}
+
 	// Add query params to the URL, if applicable
 	let url = PROGRESS_API_ROUTE
-	const qs = queryString.stringify({ tutorialIds, collectionIds })
+	const qs = queryString.stringify(
+		{ tutorialIds, collectionIds },
+		{ arrayFormat: 'comma' }
```
Comment (zchsh) [inline src/lib/learn-client/api/progress/get-all-progress.ts:49]: Fixes an issue with `queryString` where array params would be represented as multiple duplicate keys in the query string, which our API doesn't like. `comma` option makes em comma separated instead.

**Comment by zchsh** [comment]
@ashleemboyer @BRKalow @kendallstrautman This PR implements progress query batching, to avoid many individual requests to the Learn API.

I haven't ironed everything out here yet (see note below), but this felt ready for initial review at least. Will also send a quick overview video in Slack in case it's helpful 👍

> **Note**: I'm running into an issue with `400` requests, but I think it's largely tangential to the main event here, so ignoring it for now.

File: src/hooks/progress/use-tutorial-progress-mutations/index.ts (hunk 7 above)
Comment (ashleemboyer) [reply src/hooks/progress/use-tutorial-progress-mutations/index.ts:59]: Ready to resolve this? 🥳

File: src/lib/learn-client/api/progress/get-progress.ts [hunk 12]
```diff
@@ -26,14 +26,17 @@ type GetAllProgressResult = ApiCollectionTutorialProgress[]
  *
  * https://digital-api-specs.vercel.app/learn#tag/Progress/paths/~1progress/get
  */
-const getAllProgress = async ({
+const getProgress = async ({
```
Comment (ashleemboyer) [inline src/lib/learn-client/api/progress/get-progress.ts:29]: Appreciate the rename here! 🙌🏻

**Review by ashleemboyer — APPROVED** [review]
Went through the testing steps locally and everything works as expected! 👏🏻

**Review by kendallstrautman — COMMENTED** [review]
~~I followed the testing steps but still seeing many network requests without commenting out the `useProgressBatchQuery`. I see some 500 errors for the /progress?tutorialIds=... calls.~~ nvm! I didn't npm install 😵‍💫 my bad

### Without batch query hook
Screen Shot 2022-09-08 at 2 03 14 PM

### With batch query hook
Screen Shot 2022-09-08 at 2 03 40 PM


~~I also see the failed 500 requests continuing to fire off 🤔~~

**Comment by ashleemboyer** [comment]
@kendallstrautman I saw this too at first, and doing an `npm install` to get the latest of `query-string` worked for me. Have you already tried that?

**Comment by kendallstrautman** [comment]
@ashleemboyer ahhh okay! i'll try this 🙇

**Review by kendallstrautman — APPROVED** [review]
@zchsh sorry I messed up testing the first time. All is looking great after installing things properly, awesome improvement 💯

**Comment by zchsh** [comment]
@kendallstrautman all good 💖  I def should have specified the `npm i` in testing steps -- the fact `query-string` needed that `comma` option thing, and the fact that option wasn't available except in recent versions, threw me on a loop when working on this 😅
